pystiche_papers/sanakoyeu_et_al_2018/_modules.py

```diff
@@ -0,0 +1,186 @@
+) -> Union[Tuple[int, int], int]:
+    padding = verify_str_arg(padding, valid_args=["same", "valid"])
+    if padding == "same":
+        return cast(Tuple[int, int], same_size_padding(kernel_size))
+    else:  # padding == "valid"
+        return 0
+
+
+def get_activation(act: str = "relu", inplace: bool = True) -> nn.Module:
+    act = verify_str_arg(act, valid_args=["relu", "lrelu"])
+    if act == "relu":
+        return nn.ReLU(inplace=inplace)
+    else:  # act == "lrelu"
+        return nn.LeakyReLU(negative_slope=0.2, inplace=inplace)
+
+
+def conv(
+    in_channels: int,
+    out_channels: int,
+    kernel_size: Union[Tuple[int, int], int],
+    stride: Union[Tuple[int, int], int] = 2,
+    padding: str = "valid",
+) -> nn.Conv2d:
+    padding = get_padding(padding, kernel_size)
+    return nn.Conv2d(
+        in_channels, out_channels, kernel_size, stride=stride, padding=padding
+    )
+
+
+class ConvBlock(nn.Sequential):
+    r"""ConvBlock from :cite:`SKL+2018`.
+
+    This block comprises a convolution followed by a normalization and an optional
+    activation.
+
+    Args:
+        in_channels: Number of channels in the input.
+        out_channels: Number of channels produced by the convolution.
+        kernel_size: Size of the convolving kernel.
+        stride: Stride of the convolution. Defaults to ``1``.
+        padding: Padding of the input. It can be either ``"valid"`` for no padding or
+            ``"same"`` for padding to preserve the size. Defaults to ``"valid"``.
+        act: The activation is either ``"relu"`` for a :class:`~torch.nn.ReLU`,
+            ``"lrelu"`` for a :class:`~torch.nn.LeakyReLU` with ``slope=0.2`` or
+            ``None`` for no activation. Defaults to ``"relu"``.
+        inplace: If ``True`` perform the activation in-place.
+
+    The parameters ``kernel_size`` and ``stride`` can either be:
+    * a single :class:`int` – in which case the same value is used for the height and
+      width dimension
+    * a tuple of two :class:`int` s – in which case, the first int is used for the
+      vertical dimension, and the second int for the horizontal dimension
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        kernel_size: Union[Tuple[int, int], int],
+        stride: Union[Tuple[int, int], int] = 1,
+        padding: str = "valid",
+        act: Union[str, None] = "relu",
+        inplace: bool = True,
+    ) -> None:
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+        modules: List[nn.Module] = []
+        modules.append(
+            conv(in_channels, out_channels, kernel_size, stride=stride, padding=padding)
+        )
+
+        modules.append(nn.InstanceNorm2d(out_channels))
+
+        if act is not None:
+            modules.append(get_activation(act=act, inplace=inplace))
+
+        super().__init__(*modules)
+
+
+class DeConvBlock(nn.Module):
```

Comment:
In the paper they call it "upsampling block". Do they use `DeConv` in the implementation? If not, I suggest we use `UpsampleConvBlock`.

Reply:
They call it [deconv2d](https://github.com/CompVis/adaptive-style-transfer/blob/07a3b3fcb2eeed2bf9a22a9de59c0aea7de44181/ops.py#L58) in the implementation.

Reply:
The problem I see is that de-convolution is an [existing term](https://en.wikipedia.org/wiki/Deconvolution) that basically describes the inverse operation of a given convolution. Although the authors used that in their reference implementation, I think we should not use this to avoid confusion.